### ISSUE TITLE
Add ability to extend from other settings files

### ISSUE THREAD

**Issue by maxdeviant** [Microsoft/vscode#15909]
Inspired by TSLint's ability to [extend](https://palantir.github.io/tslint/2016/03/31/sharable-configurations-rules.html) configuration files, it would be nice if `.vscode/settings.json` could behave the same way.

So if I have some global settings set up:

`~/example-repo/.vscode/my-company-settings.json`:
```json
{
  "editor.tabSize": 2,
  "editor.insertSpaces": true,
  "editor.formatOnSave": false
}
```

I can use them in another file, without having to duplicate the settings:

`~/example-repo/my-project/.vscode/settings.json`:
```json
{
  "extends": "../../.vscode/my-company-settings.json",
  "editor.formatOnSave": true,
  "editor.fontLigatures": true
}
```

And the computed settings for `~/example-repo/my-project/.vscode/settings.json` would be:

```json
{
  "editor.tabSize": 2,
  "editor.insertSpaces": true,
  "editor.formatOnSave": true,
  "editor.fontLigatures": true
}
```

Scenario:

Multi-root workspace doesn't solve this for our use case. We have a bunch of npm modules each in their own git repository. We have a package which contains our shared tsconfig.json and tslint.json settings that all the other packages include with extends. We don't use a multi-root workspace since the idea is that people can clone the specific package(s) they need to work on. Every repository contains the exact same .vscode directory which is essentially copy&pasted all over the place. Maintaining the .vscode settings for the projects is "not pretty" compared to the tsconfig.json and tslint.json which only require the settings package to be updated with for example yarn upgrade.

**Comment by mweststrate** [comment]
+1: Our mono repo consists of six projects which have largely, but not entirely the same `settings.json` files

**Comment by maxdeviant** [comment]
@mweststrate A monorepo is exactly why we desire this as well :smile:

**Comment by andreineculau** [comment]
@sandy081 am I correct in saying that you're advocating that large groups that share common settings should create their own extension? https://github.com/Microsoft/vscode/issues/15162#issuecomment-262449840

If you know of such an extension that would be a great reference. Thanks!

**Comment by sandy081** [comment]
@andreineculau No, I did not mean that. I was mentioning about settings contributed by extensions.

**Comment by mmkal** [comment]
@bpasero @jrieken any update on this? Would the team be open to PRs?

IMO the lack of this feature is the reason that questions like [should I commit the .vscode folder to source control?](https://stackoverflow.com/questions/32964920/should-i-commit-the-vscode-folder-to-source-control) are controversial/asked at all. This isn't just useful for monorepos - it could allow having settings specific to a project and user.

Say you have team settings checked in to `.vscode/settings.json`, but you want to set something to some specific value, only for a single project. Currently you'd have to make that change in the team `.vscode/settings.json` file, which sucks because you'd the either have to check it in to source control for the whole team, or add the file to `.git/info/exclude` (in which case every time you _do_ need to update the file in source control, you have to  undo the exclusion, remove your personal settings, apply the team changes, check-in, re-apply your personal settings and redo the exclusion).

If settings/tasks/launch.json supported `extends`, you could have a `.vscode/settings.team.json` with the team settings, then `.vscode/settings.json` could be:
```json
{
    "extends": ["./settings.team.json", "./settings.user.json"]
}
```
where `.vscode/settings.user.json` is in `.gitignore`. `launch.json` and `tasks.json` could work in exactly the same way (I guess this feature would probably need to be implemented so it can handle the `./settings.user.json` not existing).

**Comment by bpasero** [comment]
@mmkal with multi root workspaces we have a new way of defining settings without having to check them in. Currently only available in insiders (https://code.visualstudio.com/insiders/) you can save a workspace with any number of folders (File > Save Workspace As) and this will create a `*.code-workspace` file where you can add workspace settings into. All settings are scoped to this workspace and they will not get checked in to SCM.

The only downside of this approach is that settings defined within the folders will still override the workspace settings. But I think that is fair because a setting that is checked into SCM should always have higher importance imho.

**Comment by amp343** [comment]
@bpasero that feature doesn't really address the use case that @mmkal described though, where a base `settings.json` (presumably tracked) would be extendable/overridable by a `settings.user.json` (presumably untracked).

The use case for this is that a team may want to track and distribute shared settings for a folder (using the existing `settings.json`), but give individual users the flexibility to opt out of those settings (via `settings.user.json`).

The pattern I'm thinking of is similar to the one used in `docker-compose`, (https://docs.docker.com/compose/extends/) where you might have a tracked `docker-compose.yml` (shared base/prod config), and an untracked `docker-compose.override.yml` (local development overrides) that automatically extends from the base config.

I think something like this could be a powerful feature, especially for adoption across larger teams.

**Comment by mmkal** [comment]
@bpasero @amp343 is right, multi-root workspaces are great, but they only really solve problems to do with having related projects, not so much version control/different developers working on a repo - or tasks/launch.json. 

One approach that would solve for almost all use cases would be to allow `tasks.js`, `launch.js` and `settings.js`, as well as `.json` files in the `.vscode` folder. That would allow for this, and many other features too. Similar to how jest allows defining a `jest.config.js` file where you just put `module.exports = { ... }`.

That way if a team needed to have default settings, overridable by individual developers locally they could define whatever custom extension logic they liked in`settings.js`:

```js
const { existsSync } from 'fs'
const localPath = __dirname + '/settings.local.js'
const userOverrides = existsSync(localPath) ? require(localPath) : {}
module.exports = {
  // team settings
   'tslint.enable': true,
   ‎...userOverrides,
}
```

If you wanted to get fancy, you could use `lodash.defaultsDeep`. For very complex or widely-used configs, people could even publish their configs to npm and they could be intalled as npm dev dependencies, and used by others with very little effort.

Again, would the team be happy to accept PRs for something like this? It seems like multiple people would be willing to help, (including me).

**Comment by ilius** [comment]
I also want to sync my vscode config directory (including settings.json, but excluding some specific paths like Cache*) between home and work, but there are few things I don't want to sync (like `window.zoomLevel` which i change depending on display size). If we could import / extend another json file in settings.json, that would solve my problem.

**Comment by aleksey-bykov** [comment]
at very least you can keep looking for settings in parent folders, can't you?

**Comment by sandy081** [comment]
@maxdeviant What is your use case here to want the settings to be extended? Is it to have non-sharable workspace settings or something different. 


@mmkal @amp343 To have non-sharable workspace settings did you try following?

- Open empty window
- Use command `Add folder to Workspace` and add the folder in which you want to have non-sharable settings
- Open Workspace settings and have your personalised settings for that workspace

In this way you do not need to change the team settings in the folder which are shared.

**Comment by maxdeviant** [comment]
@sandy081 Wow, it's been a while since I opened this issue, I'll see if I came remember 😅 

The original use case was that I wanted to be able to share common settings across multiple VS Code projects. So I wanted the ability to have a base `.vscode/settings.json` somewhere in our repo with common options (e.g., indent size) and then point all the other settings files to that base file and be able to inherit the base settings.

We're using multi-root workspaces now, so I don't really think this is an issue for us anymore.

**Comment by sandy081** [comment]
Closing this as the multi-root workspaces solve the issue. Others please use https://github.com/Microsoft/vscode/issues/40233 for non sharable workspace settings.

**Comment by qtiki** [comment]
Multi-root workspace doesn't solve this for our use case. We have a bunch of npm modules each in their own git repository. We have a package which contains our shared `tsconfig.json` and `tslint.json` settings that all the other packages include with `extends`. We don't use a multi-root workspace since the idea is that people can clone the specific package(s) they need to work on. _Every_ repository contains the exact same `.vscode` directory which is essentially copy&pasted all over the place. Maintaining the `.vscode` settings for the projects is "not pretty" compared to the `tsconfig.json` and `tslint.json` which only require the settings package to be updated with for example `yarn upgrade`.

**Comment by sandy081** [comment]
@qtiki Makes sense. Reopening for your scenario and updating it in the description

**Comment by DJMcNab** [comment]
I now dual-boot Windows and Ubuntu, and was looking for a way to share my settings between them. This would be my use case for a feature like this.

**Comment by haugerbr** [comment]
@DJMcNab you might be better off with something like this: https://marketplace.visualstudio.com/items?itemName=Shan.code-settings-sync

**Comment by huan** [comment]
+1

Should be just like the new TypeScript 3.2 feature: [tsconfig.json inheritance via Node.js packages](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#tsconfigjson-inheritance-via-nodejs-packages)


> TypeScript 3.2 now resolves tsconfig.jsons from node_modules. When using a bare path for the "extends" field in tsconfig.json, TypeScript will dive into node_modules packages for us.
> 
> ```json
> {
>     "extends": "@my-team/tsconfig-base",
>     "include": ["./**/*"]
>     "compilerOptions": {
>         // Override certain options on a project-by-project basis.
>         "strictBindCallApply": false,
>     }
> }
> ```
> 
> Here, TypeScript will climb up node_modules folders looking for a @my-team/tsconfig-base package. For each of those packages, TypeScript will first check whether package.json contains a "tsconfig" field, and if it does, TypeScript will try to load a configuration file from that field. If neither exists, TypeScript will try to read from a tsconfig.json at the root. This is similar to the lookup process for .js files in packages that Node uses, and the .d.ts lookup process that TypeScript already uses.
> 
> This feature can be extremely useful for bigger organizations, or projects with lots of distributed dependencies.
>

**Comment by superole** [comment]
> at very least you can keep looking for settings in parent folders, can't you?

That would solve one use case that I see for this feature: -The ability to open Vs Code in a subfolder of a parent project and still use the launch/settings/tasks of the parent project.

**Comment by stiltet** [comment]
I 100% agree with @superole

**Comment by joem-msft** [comment]
My team has an Open Source repo that contains many separate projects. Each project is usually independent from all other projects. We want to make it so that each project can be opened alone (single-folder workspace) yet they should share the same base settings as all other projects in the repo. We can see there needing to be the following setting files:

```javascript
// {repoRoot}/configs/repo-settings.json
{
    "editor.insertSpaces": true,
    "[json]": {
        "editor.detectIndentation": false,
        "editor.tabSize": 2,
        "files.trimTrailingWhitespace": true
    }
}
```
```javascript
// {repoRoot}/configs/projects-settings.json
{
    "$extends": "./repo-settings.json",
    "json.schemas": [
        {
            "fileMatch": [
                "/apiDefinition.swagger.json"
            ],
            "url": "http://json.schemastore.org/swagger-2.0"
        }
    ]
}
```
```javascript
// {repoRoot}/configs/samples-settings.json
{
    "$extends": "./repo-settings.json"
}
```
And some shared tasks:
```javascript
// {repoRoot}/configs/projects-tasks.json
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "build",
            "command": "dotnet",
            "type": "process",
            "group": {
                "kind": "build",
                "isDefault": true
            },
            // and others
            "options": {
                "cwd": "${workspaceFolder}/FunctionApp"
            }
        }
    //
    ]
}
```

And then in each project, we could simply have the following as the default template
```javascript
// projects/MyProjectA/.vscode/settings.json
{
    "$extends": "../../../configs/projects-settings.json"
    // Custom project-specific overrides would go here
}
```
```javascript
// projects/MyProjectA/.vscode/tasks.json
{
    "$extends": "../../../configs/projects-tasks.json"
    // Custom project-specific overrides would go here
}
```

And the same similar template for sample projects, but they'd be able to choose to extend a different settings file.

This would require:
1. explicit declaration of the settings file to include, using paths relative to the current settings file.
1.a. Feel free to choose some property name for this. I've used "$extends" in the examples above.
2. recursive extending.
2.a. Would need to detect circular extends.
3. Allow this to work with launch.json, tasks.json, etc.
4. Resolving of variables like {workspaceFolder} should be resolved AFTER it's included via an $extends.
5. The originating setting file would determine the precedence (e.g. workspace vs user, vs folder). Simply treat the expanded settings.json file in the place of that scope.

**Comment by peredin** [comment]
This would be a nice feature, I'd like it myself for sharing a base vscode configuration across my own computers.

**Comment by paparent** [comment]
I'm quite interested in that feature!

Yes, having ability to extends from a local setting that is git ignored that would be nice.

In a mono repo too, with workspaces, so we can have base settings such as linters config and such.

**Comment by mchelen-gov** [comment]
wanted to highlight for folks:
- [workaround](https://github.com/microsoft/vscode/issues/15909#issuecomment-434604391)
- https://github.com/microsoft/vscode/issues/40233 non-sharable workspace settings issue

**Comment by ladjzero** [comment]
It would be great to share workspace settings between micro-service projects.

**Comment by RobertLowe** [comment]
`extends` seems like an elegant and simple solution

**Comment by hrobertson** [comment]
In my case I'm not currently needing to share settings between repos, just wanting to have a `settings.json/js` that is checked in and allowing for that to be overridden in a `settings.local.json/js` which is git ignored.

**Comment by lukaszpolowczyk** [comment]
The variable `$ {commonWorkspacesSettingsFolder}` will also be useful, which is the address to a special folder "(...)/Code/CommonWorkspaceSettings" folder in User Settings. (final name, of course, to be determined)

This folder should contain settings that are universal for many projects.

This is my suggestion from my duplicate: https://github.com/microsoft/vscode/issues/38162

**Comment by jacobisaliveandwell** [comment]
@haugerbr Hello. I see that you're the most recent member to comment.

Similar to `.vscode/extensions.json` and how it's safe to commit `recommendations` to a VCS, because even per-workspace extension settings don't modify that file, it would be nice to have a file with sane defaults that are repo/team specific, and are overridden by an individual's `settings.json`.

I suggest something like `shared-settings.json`, but I think the OP was shooting for a more generalized solution that could also fit multi-root workspaces. For that, I vote for [this comment](https://github.com/microsoft/vscode/issues/15909#issuecomment-540540211).

An example use case is repos with complex extension configurations, like Puppet which offers `modulePath`, and every developer would need to set this manually, or deal with VCS conflicts, or manage settings globally, even though they might be workspace specific. Without having a means of setting per-workspace "shared" or "extended" settings is leading to a poor user experience.

I hope this all makes sense. Let me know if I can elaborate.

**Comment by haugerbr** [comment]
@jacobisaliveandwell Unfortunately, I don't work on VS code. The member tag is applied to anyone that is a part of the Microsoft organization. You can tag sandy081 who does work on VS code as far as I know.

**Comment by jacobisaliveandwell** [comment]
@haugerbr Thanks!!

@sandy081 Think you could share your insights on getting this some attention? Thanks.

**Comment by sandy081** [comment]
Unfortunately no plans at present to implement this. Sorry for saying so.

**Comment by jacobisaliveandwell** [comment]
This makes me a sad panda

**Comment by octogonz** [comment]
> Unfortunately no plans at present to implement this. Sorry for saying so.

@sandy081 If someone from the community did the work, would you accept a PR to implement this?

Large monorepos can have hundreds of projects. Today I saw someone copy+pasting a block of settings into the `.vscode/tasks.json` for many different projects -- that is going to be difficult to maintain.

**Comment by JoshuaKGoldberg** [comment]
@sandy081 would you be open to an external contributor _(me!)_ sending a PR for this feature?

My team at work manages several repositories of various sizes and we'd like to standardize a few features around them so we don't have to keep updating checked-in `.vscode/settings.json` files for each update.

**Comment by sandy081** [comment]
Sure, but please draft the approach/design first so that I can review with team.

**Comment by JoshuaKGoldberg** [comment]
Super. @octogonz you did mention interest a while ago; are you still up for it?

**Comment by globalhuman** [comment]
My team are currently using a system where we having a `settings.json.template` which is committed to our repository. Unfortunately we have to manually update our `settings.json` when modifications are made or the repo is pulled fresh.

Having the ability to extend `settings.json` as is already done by [snippets](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_project-snippet-scope) with a `name.code-snippets` file in the `.vscode/` directory.

Hopefully the `vscode` devs can take a look at what's already been done there and add this much needed functionality

**Comment by FrederickEngelhardt** [comment]
One workaround solution could be writting a bash file that takes a base vscode settings and then recursively merges the .json files based on order.

Example peudo code

files
0-base.json
1-repo.json
2-user.json

Then run a JS or shell script post install and create a `.vscode/settings.json` file for the merge config. The `settings.json` would be ignored in git and maybe anything with the starting characters of `2-*` would be ignored in settings.json.

Would only be worth doing this for monorepos that require different configs + user's requiring custom local config.

Anyone already do this?

**Comment by jstevens-level20** [comment]
Having the ability to use multiple settings files would be awesome.  We have some settings we want to retain in our repos (e.g. import sorter), but others that we don't (e.g. peacock)

**Comment by haugerbr** [comment]
@jstevens-level20 isn't that already solved by [settings precedence](https://code.visualstudio.com/docs/getstarted/settings#_settings-precedence). I would assume peacock would be in someone's user settings which are always stored locally.

**Comment by RobertLowe** [comment]
@haugerbr not really when sharing with other developers or multiple repos, what if I want to extend from my organizations settings, its just a good way to organize settings

**Comment by RobertLowe** [comment]
Maybe this can be solved by allowing for `js` settings files, like eslint/babel/etc does this way a developer an implement whatever setting resolution they want all from a simple `require` from vscode's side...

**Comment by haugerbr** [comment]
@RobertLowe I think that this issue was created for the situation where you have multiple settings/workspaces in a repo and they have a lot of duplication so you want to have a core settings file in the repo that you can extend from.

For the situation you discussed, if I open a workspace with settings in a repo, those settings get applied on top of my user settings, essentially extending from my user settings already. I'm unsure in what situation you would need an explicit extend between user and repo settings.

I'm mainly saying this because it sounds like what @jstevens-level20 wanted is solved by settings precedence and the multiple settings files that already can exist.

**Comment by haugerbr** [comment]
@RobertLowe I guess I see your point if you want user settings that you don't want to set up that someone else created but I was really only trying to understand why @jstevens-level20 issue isn't solved by settings precedence. I don't work on the VSCode team myself but like make sure people aren't waiting on something that may already be solved in another way.

**Comment by NonLogicalDev** [comment]
I would like to add another usecase where user/project settings distinction is not enough.

I have often hop between systems, Linux / Mac / Windows, work and home. And most of the VSCode settings apply across the board, but there is a small set of system local settings. Like locations of binaries for certain plugins or other sorts of location / system dependent configurations. I keep my configuration checked into a git repo along with configuration for other development tools.

At the moment the best I can do is set up default settings and just ignore the VSCode settings file which gets populated with system specific stuff.

A better approach is to have a (local system specific) file that imports a bunch of global configurations (which can be tracked in a repo). That way at home you can import your global settings, and at work you can import your settings and additionally overload work provided ones. And overload the local system specific setting in yet another file.

**Comment by danielcompton** [comment]
Emacs has a clever solution to this:

> You can also use `.dir-locals-2.el`; if found, Emacs loads it in addition to `.dir-locals.el`. This is useful when `.dir-locals.el` is under version control in a shared repository and can’t be used for personal customizations.

The advantage of this method is that you wouldn't even need to touch the `settings.json` file to add your own local customisations. You would add `settings2.json` to your `.gitignore`, and VSCode would merge `settings.json` and `settings2.json`.

**Comment by jacob-israel-turner** [comment]
> The advantage of this method is that you wouldn't even need to touch the `settings.json` file to add your own local customisations. You would add `settings2.json` to your `.gitignore`, and VSCode would merge `settings.json` and `settings2.json`.

I love this approach.  A more robust inheritance/precedence model should mean nobody needs to change any configuration to benefit - a developer joining a project would just need to add something like `settings2.json` or `settings.user.json` in order to override workspace settings.

**Comment by effs** [comment]
This is also a privacy concern: I may not want certain non-secret but private or confidential info to leak through settings.

In short, there's a huge number of use cases for extensible settings files.

A naive solution would be an ordered list of settings files, each of which overrides the previous. The implementation is pretty much `Object.assign`.

**Comment by FlorianWendelborn** [comment]
@effs for some settings, a deep assign also makes sense though. For example, you may want to add more experimental file nesting settings rather than replacing them all

**Comment by adam-coster** [comment]
I was also hoping for something like this, via something like an `"extends": ["../settings.json", "./other-settings.json"]"`. Collections of settings I want to be able to independently mix & match include:

- File folding settings
- Search-ignored patterns
- Typescript settings

I also want to be able to have *multiple* multi-root workspace configurations in a single repo, so that I can work on subsets of highly-related projects with minimal cognitive and performance overhead. Those will tend to use nearly the exact same base settings, since all of my projects use the same tech, build processes, etc. Since I cannot extend from base settings, having multiple `.code-workspace` configs requires basically copy-pasting the entire `"settings"` field between them. And if I want to add or change a setting in one, I probably also want to in the other one, and so maintaining consistency is difficult.

This is frustrating enough that I'll likely be running a build process to assemble my `.code-workspace` files from multiple `settings.json` files!

<details>
<summary>I've provided a toy example in case all that was unclear</summary>

For a small, **toy example** git repo with the following content:

```
# Three highly-related projects collectively making up a website
projects/website-server
projects/website-docs
projects/website-front

# Other projects that share settings, build processes, etc
# but otherwise are loosely or not related to the website,
# along with repo-wide configs and other info.
projects/some-product
projects/tools
package.json
.env

# VSCode folder containing workspace configs for
# two workspaces, each of which extend from the shared settings.
.vscode/settings.shared.json
.vscode/everything.code-workspace
.vscode/website.code-workspace
```

All of which are Node + Typescript projects, I'd have three different workspace configurations that can all build upon a core set of shared settings:

```json
// everything.code-workspace
{
	"folders": [ // All the project folders, plus the root
	],
    "settings": {
		"extends": ["./settings.shared.json"],
		// ... other settings when using this multi-root workspace to see
       // all projects at the same time ....
	}
}
```

```json
// website.code-workspace
{
	"folders": [ 
		{"path": "../projects/website-server"},
		{"path": "../projects/website-docs"},
		{"path": "../projects/website-front"}
	],
    "settings": { 
		"extends": ["./settings.shared.json"],
		// ... settings that make sense when just looking at these highly-related projects
	}
}
```

(Even if we could only extend regular `settings.json` files (versus the `"settings"` field inside a workspace config) that would allow for extremely minimal settings configurations in each workspace folder's `.vscode/settings.json` file.)

</details>

**Comment by effs** [comment]
> @effs for some settings, a deep assign also makes sense though. For example, you may want to add more experimental file nesting settings rather than replacing them all

Oh for sure. There are any number of deep merge implementations too (probably want to avoid jsonpath etc.) but really point was just that this should be relatively straightforward for the 80% case. There are a number of corner cases, I'm sure.

**Comment by SkyCodeSky** [comment]
Are there any updates yet?

**Comment by fritx** [comment]
@SkyCodeSky you can use tools like `jsonmerge` to deep-merge separate files into one: 

ref: https://github.com/Jayin/jsonmerge/pull/1

```sh
jsonmerge --jsonc .vscode/settings_*.json > .vscode/settings.json
```

**Comment by effs** [comment]
> @SkyCodeSky you can use tools like `jsonmerge` to deep-merge separate files into one:

Good tip, but we all manage _somehow_ :) There’s even a few extensions that do some degree of merging. It’d just be nice to have support for extending built in so that we don't have to cobble almost-but-not-quite solutions together.

**Comment by bobvanderlinden** [comment]
I use home-manager to manage most of my home-directory configuration files. For vscode, there are a number of settings I'd like to manage by home-manager to make them aligned with other applications (like fonts). Some settings I'd like to manage from vscode (like the custom dictionary words).

Having 2 files enables home-manager to manage one of them, but have the other be mutable from vscode itself. This feature would be really helpful for that.

**Comment by foxx** [comment]
I would be strongly in favour of such a feature, but here's a workaround that may help others.

Our use case is approx 300+ repos, each configured for devcontainers with a custom image, and primarily using the same set of languages throughout (Rust, Python, TypeScript). There's a long list of entries for `settings.json` that repeated across all repos, such as `files.exclude` for `**/__pycache__` etc.

Extendability for `settings.json` is typically an opinionated operation, meaning you have to define a merge strategy, as seen in tools such as `jsonmerge` and decide what order to load the files.

For ordering, you can enforce this by adding numbers at the start of the file for glob style ordering:

```
01_settings_python.json
02_settings_rust.json
03_settings_general.json
```

You could then have a post-processing script in every repo that will do the following:

* Merge all JSON files in `.vscode/settings/` using a chosen JSON merge strategy
* This could perhaps even be defined as a special `__strategy__` key in the top-level JSON, which you could remove during processing.
* Another special operation could be to source the settings from another location (such as HTTP or disk etc).
* Output resulting JSON into `.vscode/settings.json`

Ultimately this would require a script to be maintained, so our approach will likely be to put this script into a custom devcontainer image which is shared by all repos. For example `generate-vscode-settings [path-to-vscode-folder]` or something like that.

This is a hacky workaround, but it sure beats maintaining 300+ copies of `settings.json` by hand!

**Comment by sanmai-NL** [comment]
@foxx: you can add a Git submodule with a default `.vscode/settings.json`, and symlink to it from each repository's root directory. It works pretty well across standard products, e.g., a Python app, a Rust library, etc.

I do that, but it falls apart when you need to extend from a base config, it falls apart. For example, I set a different activity bar background color (`activityBar.background`) for some major repositories that we juggle a lot concurrently. This color must  not be the same one or more repositories.

**Comment by nolimitdev** [comment]
@maxdeviant you can easily hardlink your file `my-company-settings.json` to as many locations as you need

**Comment by starball5** [comment]
Related Stack Overflow question: [Monorepo - Multiple .vscode settings](https://stackoverflow.com/q/72808674/11107541)

**Comment by movahhedi** [comment]
Adding the ability to create different setting files such as `settings.js`, `settings.cjs`, `settings.mjs` can automatically solve this issue, by importing and exporting another settings file. I use this approach to use shared configs in Prettier, ESLint, StyleLint, CSpell. I'm surprised that vscode still doesn't support js config files.

```js
module.exports = {
	...require("../.eslintrc.json"),
};
```

**Comment by RobertAKARobin** [comment]
This issue is especially important with the release of profiles. It's great to be able to have different profiles for different types of projects. However, there are many things that we want to keep consistent between all our projects. We were surprised that even if a profile's `settings.json` is blank, most of the settings in the root `settings.json` are disabled completely.

**Comment by aaronsteers** [comment]
I came here for the same reason @RobertAKARobin notes above:

> This issue is especially important with the release of profiles.

I'll spin off a side issue for the Profiles use case specifically...

**Comment by birgersp** [comment]
our team has a monorepo that is split into mulitple workspaces. sharing settings (about 100+ lines now) between the workspaces is a PITA. please implement this issue.

our only way of doing this now is to use our own "workspace-settings-sync" tool.

**Comment by matepek** [comment]
VSCode developers intention was that `.vscode/*` files are not shared, local-only thing. (should be in `.gitignore`).
Shared settings should go into the `project.code-workspace` file and the precedence is `.vscode/*` > `project.code-workspace` > "user settings" > "default settings". 
See [details here](https://code.visualstudio.com/docs/getstarted/settings#_settings-precedence).

**Comment by starball5** [comment]
> VSCode developers intention was that .vscode/* files are not shared, local-only thing. (should be in .gitignore).

@matepek Um... what? Quoting from [the very page you linked](https://code.visualstudio.com/docs/getstarted/settings#_workspace-settings):

> Workspace settings are specific to a project and can be shared across developers on a project. Workspace settings override user settings.

**Comment by diminutivesloop** [comment]
Yeah @matepek I think you're mistaken. If the `.vscode` folder isn't supposed to be committed [that must be news to the VSCode team](https://github.com/microsoft/vscode/tree/main/.vscode).

**Comment by ajlive** [comment]
> I'm interested of this feature for different reason, just posting this because I would like to remind some newcomer that their use case is already supported.

Not exactly a newcomer to vscode and what I want is not covered by workplace settings. When I say I want to share my settings.json, I don’t mean for a particular project. I’ve had people ask me advice about vscode settings and want to be able to just send them a link to my settings in my dotfiles repo, as I do with vim, eMacs, and shell config. I ended up writing a hacky Lua script to put things together. I looked into using existing tools, but afaict the ones that can handle jsonc strip out the comments when merging.

**Comment by trailstrider** [comment]
I don't need to be able to extend settings per se, but I do need to be able to separate settings for the folder/workspace into those that get committed to the repo and those that do not.  

The challenge in implementing just this is that it is up to the users to decide what goes in .gitignore... but in typical VS Code "nag-the-dev-with-toasts" style, it could warn when the file that isn't supposed to be committed by convention is missing from .gitignore.   

But otherwise, when I came across this thread discussing extending, it made a lot of sense to put it on the users to extend and put stuff in ignore if they want.... The only issue then is that we'd really need the default save location for workspace/folder settings should be in the settings file that gets ignored.  But without a default intended to be ignored and one intended to be shared, how can VS Code begin to know where new folder/workspace settings should go?

This brings me back to wanting VS Code to have a default pair of files (.vscode/user_settings.json and a .vscode/shared_settings.json perhaps?), independent of the ability to extend.

**Comment by bradzacher** [comment]
At Canva we have a large repo and users want to add their own custom, local settings for the repo as they configure various plugins specifically for the company's workflows.

They can't use user settings for this usecase because much of the config is required specifically for the company's setup - so they don't want to leak it.

We also want to commit a `settings.json` to the repo because we want to provide and control a set of "sensible defaults" for the repo to help users get started and stay in sync as we make changes to the infra over time.

**Comment by vhlushch** [comment]
Hi @sandy081, hope you have a great day. Will this be prioritized, considering amount of upvotes (620 as of today)?

**Comment by muchisx** [comment]
@sandy081 I think you got enough "use case" explanations already, any plans to move forward with this? 👋🏻

**Comment by sscaff1** [comment]
This would be incredibly useful.

**Comment by nolimitdev** [comment]
> Would love to have this. I need to split my workspace settings into two: One file that I can commit and share with coworkers, and one for personal customization.

We exactly do this using this way: https://github.com/microsoft/vscode/issues/15909#issuecomment-1410351025 ...

> We partially solved it by sharing team settings in commited file .vscode/project.code-workspace and if developer need some project-specific settings just for himself he can use gitignored .vscode/settings.json which has precedence. I think that better fallback does not exist now.

Please try it and give feedback!

**Comment by ackvf** [comment]
Hello everyone, we have a shared `.vscode/settings.json` file checked in repo and I would like to add some settings that are unrelated to the project in the sense that they should not be checked into the repo as they are a mere personal preference. How can I achieve it? I only want to make this change for this project, not my whole VSCode.

My idea was to use `.vscode/settings.local.json` and put it in my _global_ git-ignore file but that doesn't seem to work.
As a workaround, is it possible to put it under a scope in my vscode user-settings?

Cheers 🥂

**Comment by starball5** [comment]
@ackvf read literally the comment directly above yours :|

**Comment by RobertLbebber** [comment]
Unless I'm not getting something, which is likely the case. There is a big distinction between workspace and setting.json.
Why would I want to commit my color theming to a repo in the .workspace, yet I can't set that in the settings.json. That flies in the face of what should be committed versus what shouldn't be. Another one is having the side bar on the left rather then right, this would have to be commited in the workspace file rather than the personal settings file

**Comment by nolimitdev** [comment]
> @Andonvr
> 
> > Would love to have this. I need to split my workspace settings into two: One file that I can commit and share with coworkers, and one for personal customization.
> 
> We exactly do this using this way: [#15909 (comment)](https://github.com/microsoft/vscode/issues/15909#issuecomment-1410351025) ...
> 
> > We partially solved it by sharing team settings in commited file .vscode/project.code-workspace and if developer need some project-specific settings just for himself he can use gitignored .vscode/settings.json which has precedence. I think that better fallback does not exist now.
> 
> Please try it and give feedback!

This is not ideal solution as some use-cases will not work. It is just fallback for missing official feature of extending settings. For example if I have `{ "settings": { "window.title": "Foo" } }` in `./vscode/project.code-workspace` and then when I try override it to custom developer value using `{ "window.title": "Bar" }` in `./vscode/setting.json` it does not work: `This setting cannot be applied in this workspace. It will be applied when you open the containing workspace folder directly.`. But for example overriding `stylelint` setting works this way so we are happy for that partially working fallback. Everybody should test its use-case if it is possible override what he need.

**Comment by nbbaier** [comment]
I would also love this feature, adding my vote.

**Comment by nevercast** [comment]
Even without the `"extends"` functionality, we would love to have `settings.user.json`. We have team configuration in `settings.json`, which is tracked in Git. This helps our team have a common configuration for extensions, linting etc. 

But some of our team members also want to make stylistic choices in the workspace, such as `workbench.colorCustomizations`. Particularly on a per-project basis as it helps them context switch when moving between projects (workspaces). 

Currently we can't offer such changes without:
1. Asking them to apply that setting at a user level and apply their customisations to the entire VSCode experience (which gets overridden by any team settings in .vscode/settings.json anyway.) This still doesn't give them per workspace overrides.

2. Making settings.json assumed unchanged in Git, then the user can apply their changes on top of the team settings. Which is more technical than we would like and not something we want our teams to be doing. Ideally.

It seems this issue has been raised in the past, e.g. #38902 which was closed for not being on the roadmap. That was almost 6 years ago. This issue itself is 7 years old.

It would be fantastic to get some feedback from the VSCode team on if they are willing to put this on their roadmap. Further, if they are willing to accept PRs to support either the `settings.user.json` functionality, or the `"extends"` functionality.

**Comment by birgersp** [comment]
@nevercast 
> ... support either the `settings.user.json` functionality, or the `"extends"` functionality.

I'm curious, why would you possibly want `settings.user.json` feature, over "extends" functionality? The latter already covers the need for the first, and would be so much more flexible.

Let's focus. We need ability to extend from other settings files, not some quasi-solution that would help in just _some_ of the cases where a "extends" feature is desired.

I'll also add that I think **all** config files in VSCode should be able to "extend" from other config files. Not just settings.json but tasks.json, launch.json, workspace files (project.code-workspace), even keybindings.json. Currently, managing multiple projects in a monorepo by copy-pasting the settings between the config files is a pain in the _behind_.

Chop, chop 🔥

**Comment by nevercast** [comment]
@birgersp Because either option satisfies my requirements. 

Undoubtedly the extends feature is more versatile.

**Comment by nsgundy** [comment]
For us, we have both workflow requirements reiterated by @birgersp and @nevercast.

1. A mono-repo with multiple projects that due to their similarity share a lot of their settings. For example:
    - recommending the same extensions (we currently do that mostly in `.vscode-workspace` files)
    - common settings either via `settings.json` or `.vscode-workspace` for the extensions and the languages used in those projects
    - common tasks via `tasks.json`, like cleaning, building, running unit-tests that only differ slightly in their name to include the project label, so that when you open a workspace at the root level, its easier to tell these apart
    - common launch configurations via `launch.json`, often referencing pre-launch tasks from `tasks.json`
2. Users requiring to set some specific variables in `settings.json` that are used by common tasks and launch configurations, but are unique to their setup, like `targetIp` for the address of the system used for debugging. All too often these changes then get accidentally committed.

In short, for us, a generic `"extends"` feature that spans all configuration `.json` files would be the most useful. It would make maintaining / setting up new projects a lot easier.

By the way @nevercast, we have had the same issue around workspace colors and eventually settled on pre-defining those in the committed `.vscode-workspace` files. This has the benefit that our developers feel right at home when pair-programming. Obviously not good for those who prefer to go their own way.

**Comment by starball5** [comment]
FYI people, as a workaround to some of these use cases, you could consider using profiles. Profiles can have settings.json, tasks.json, etc. and then you can either individually select the profile for each workspace you want it to apply to, and VS Code will remember your profile selection for that workspace, or you have open VS Code via CLI with `--profile <your profile name>`.

**Comment by ackvf** [comment]
@starball5 
> @ackvf read literally the comment directly above yours :|

OMG! thanks! Though, funnily enough, or sadly, we have **both of the files** checked in the repository 🤦‍♂️🤦‍♂️, but this is a great direction anyway, thanks for the hint. I will propose consolidating them


@nolimitdev 
> We partially solved it by sharing team settings in commited file `.vscode/project.code-workspace` and if developer need some project-specific settings just for himself he can use gitignored `.vscode/settings.json` which has precedence. I think that better fallback does not exist now.


I wish the precedence was the other way round though.
Shared `.vscode/settings.json` config checked into repository and then having as much `project.code-workspace` as necessary. Each can have slight variation depending on the task at hand.

For example: some people might prefer having another repo or a workspace folder included in IDE, some not. If the project file is checked in the repository, it removes this flexibility from you.

**Comment by recursivenomad** [comment]
Chiming in to voice that this would be great in my use-case for referencing an external `launch.json` file provided by an SDK, without having to manually copy the file into my `settings.json` or `.vscode` directory every time there's an update to the SDK.

**Comment by adrian-85** [comment]
Building on this topic, is it possible to have a global `settings.json` read from a JAR file?  We are using a framework that has a unique DSL, preventing us from leveraging common linting tools.  At least this way we could add some basic rules such as tab length, rule lines, etc as part of the JAR.

**Comment by d1onys1us** [comment]
wow im already back here again. can we plz implement 🙏

**Comment by divmgl** [comment]
This is becoming much more relevant now with monorepos.

**Comment by todaypp** [comment]
For a minimal workaround, I'm disabling all project-specific `plugins`, enabling ones related to each project only for the workspace and add those into `Workspace Settings`' recommended plugins. Likewise, I'm only defining common or fallback settings in `User Settings` and project-specific settings in `Workspace Settings`.

I've always been hoping to switch vscode envs with each project-specific profile (Node.js, Python, Java, etc.), but I guess that's all I can do for a workaround 😂

**Comment by bytebeast** [comment]
did not read the whole thread, so apologies. i did however have an opinion on the name of the key... 
{
    "include.settings": ["./settings.team.json", "./settings.user.json"]
}

**Comment by dir** [comment]
I'd love this just for cleanliness. Sometimes my settings.json files can get absolutely massive if they have large arrays, etc. Would be nice to have a linting.json, formatting.json, editor.json, etc, and then merge them all

**Comment by adam-azarchs** [comment]
Minor note, for the `settings.user.json` use case, which I think is what a lot of people are interested in, it would be most helpful if vscode would silently ignore it if the file was missing, since one would usually want to put that file in `.gitignore`.  You probably wouldn't want it to silently ignore missing files in other cases, as that could be confusing to debug, so it might be helpful to permit something like
```json
{
    "extends": [
        "settings.team.json",
        {
            "file": "settings.user.json",
            "ignore_missing": true
        }
    ]
}
```

**Comment by DeanAyalon** [comment]
Just to note, settings shouldn't necessarily be the only file extended, as extensions.json and others should probably also have this functionality

Within a monorepo, multiple projects have different recommended extensions, and opening the entire monorepo should recommend all of them

**Comment by RobertAKARobin** [comment]
Is there anything we as a community can do to help move this from the Backlog to e.g. On Deck?

**Comment by ryanjafari** [comment]
A very helpful comment, thanks @RobertAKARobin I'm guessing this portion of code is closed source?

**Comment by Kytech** [comment]
> A very helpful comment, thanks @RobertAKARobin I'm guessing this portion of code is closed source?

I'm pretty sure that the loading of settings is part of the core open-source editor as most VSCode forks have the exact same settings loading functionality. Most closed-source components of VSCode seem to be distributed as extensions (bundled in the standard VSCode distribution or otherwise), or have closed-source web backend components.

> Is there anything we as a community can do to help move this from the Backlog to e.g. On Deck?

I think the most helpful thing we could do as a community is try to come up with a proposed implementation for settings file extension, then draft a PR for the feature. Seems like there's been a number of different ideas for how this feature should be implemented, but if there's an implementation that those interested could agree upon, it would probably be relatively straightforward to implement. Having a coherent design for the feature would also make it easier for several people to contribute to the PR if one single person can't find the time to work on this. I've wanted to sit down and take a look at what would be required to implement this feature in a PR, but there's just too many other things that take priority over this for me these days.

**Comment by paskaran** [comment]
It's unclear why this issue still hasn't been addressed by VSCode, even after almost 8 years. I was searching for a solution, only to discover that none currently exists. quite sad!

**Comment by viliusan** [comment]
Bumping this issue as we are also looking for a way to extend a base `settings.json` file in a monorepo context.
It sucks to have duplicate settings for every application in a repository.

**Comment by FrederickEngelhardt** [comment]
> Bumping this issue as we are also looking for a way to extend a base `settings.json` file in a monorepo context. It sucks to have duplicate settings for every application in a repository.

@viliusan Why would you have duplicate settings per monorepo package? Each package can inherit from a base tsconfig.XYZ.json it does not need to inherit from the default file.

The only thing a monorepo package likely needs differently (and not managed by the parent) is the inclusion folder. Sometimes there are some paths that are different.

**Comment by kasvtv** [comment]
I'd love to provide a checked-in base settings file with simple defaults for the workspace, that other devs on my team can then extend using a gitignore file.

**Comment by charlietran** [comment]
This extension appears to add this capability, has anyone here tried? https://marketplace.visualstudio.com/items?itemName=KalimahApps.tabaqa

**Comment by ArseniyKruglov** [comment]
@charlietran, Tabaqa is better than nothing. Unfortunately, it can't extend multiple files.

**Comment by muzafferkadir** [comment]
# Managing Multiple Projects in VS Code with Git Integration

When working with multiple projects in a single repository, especially when the Git repo is in the parent folder, here's how to set it up in VS Code:

1. Create a workspace file (`your-project.code-workspace`) in the parent folder containing the .git directory.

2. Structure your workspace file like this:

   ```json
   {
     "folders": [
       {
         "path": ".",
         "name": "Root (Git Repository)"
       },
       {
         "path": "frontend",
         "name": "Frontend"
       },
       {
         "path": "backend",
         "name": "Backend"
       }
     ],
     "settings": {
       // Project-specific settings
     }
   }
3. You can still use .vscode/settings.json in frontend and backend folders for project-specific settings.
4. Open the workspace file in VS Code using "File > Open Workspace from File...".

**Comment by adam-azarchs** [comment]
That still doesn't solve the problem of wanting per-project configuration that is checked into git for consistency across all users on the team, e.g. `files.insertFinalNewline`, to coexist with settings which are more of a personal preference, e.g. `pylint.severity`.  You can of course just leave those settings out of `.vscode/settings.json` and set them at the workspace, machine, or user level, but it's still problematic that the `.vscode/settings.json` can't be used to provide "sensible defaults" since it overrides those other settings locations.

While I think being able to extend settings like this is very important, TBH simply being able to override the priority with which `.vscode/settings.json` was handled would solve maybe 60% of the problems I have with the current setup.

**Comment by AdoKevin** [comment]
> This extension appears to add this capability, has anyone here tried? https://marketplace.visualstudio.com/items?itemName=KalimahApps.tabaqa

This extension seems to provide a good workaround for this issue. Just need to add `.vscode/settings.json` to `.gitignore` and submit `tabaqa.json` to git. Then settings.json for each project or directory is dynamically generated by the extension based on `tabaqa.json`. The only problem is that if you are collaborating in a team, every collabrator will need to install this extension and understand how it works.

Anyway, hope that the VSCode team can provide first-party support for this.

**Comment by bravely-beep** [comment]
I'm extremely surprised that this issue has been open for almost 8 years now, and still no solution has been implemented.

Bumping this issue as I'd also find this feature very useful.

**Comment by jmevel** [comment]
@T-256 

> You can choose to either create a new profile by **COPYING** the contents from a Profile Template or from an existing profile

This is in no way an extend from another profile, it's only a simple one time copy of all settings.

**Comment by chipbite** [comment]
@T-256 : `s/inherit/copy/`

**Comment by d90mhz** [comment]
FYI, trying to rollout CoPilot across 5k+ developers... early adopters love it so far.

However, whether it's entirely true or not, people believe they need to first switch to VSCode to be effective with CoPilot.

This issue is preventing large teams from adopting. Teams working on large monorepos who are used to having others configure project setting defaults for them while still having ultimate override control for debug/diagnostic/performance reasons.

@Microsoft: you will get much better adoption of CoPilot if this trifling enhabcement is resolved.

**Comment by trailstrider** [comment]
> I might be late to the party, but given how profiles work:
> 
> It would be great if at **the very least** there would be at the **user configuration level**
> 
> 1. a key called `extendsFromDefaultProfile` that is set to false by default.
> 2. If however `extendsFromDefaultProfile==true`, your user settings inherit from the `Default` profile. Overriding any properties it defines itself. (Same merge strategy that it currently does with Project > Workspace > User, it would be Project > Workspace > User (profile) > User (default))
> 3. I see that key being only available on the default profile as one option, and another option if it would be available on all profiles (so you choose which inherit/doesnt)
> 
> Right now you either have the **exact** same settings as the default profile on another profile, or you have a COPY that you need to maintain for each profile.
> 
> The best way, of course, would be if you could `extendsProfiles: [...]` and that is an array of profiles available to the user, ignore any profile that isn't on that list.
> 
> Either way, it would allow a user to subscribe to a profile from their company/workplace/python recommendation, etc, and to be able to have their own customization to it..

It's the handling of the .vscode/settings.json being stored in the git repo for environment normalization that is at issue.  Some settings in this wind up being personal yet still unique to a particular project, while others are needed across a team so checking them in makes sense.  There isn't currently a way to differentiate the two categories of settings (shared normalization vs personal preference) at the project/workspace level.

**Comment by alexrecuenco** [comment]
> 
> It's the handling of the .vscode/settings.json being stored in the git repo for environment normalization that is at issue. Some settings in this wind up being personal yet still unique to a particular project, while others are needed across a team so checking them in makes sense. There isn't currently a way to differentiate the two categories of settings (shared normalization vs personal preference) at the project/workspace level.

No. That is not profile based, that is project based. Here we are discussing shared profiles.

**Comment by Netail** [comment]
Would be very useful for workspaces, extending from `.vscode/settings.json`

**Comment by birgersp** [comment]
> How is this not yet implemented?

Has anyone from the actual VSCode team even looked at this issue? I feel like it is just lost in their system. Like it was given to some employee who quit or something and since then nobody else knows that it exists.

**Comment by JRHeaton** [comment]
This seems like an essential feature. I just started using multiple profiles, and intuitively tried to find aa way to extend/import existing configuration(not doing files means more work ensuring state is correctly synchronized), which led me here. My only note would be that rather than "extending" being limited to a singular parent, I would value being able to "import" multiple sub-configs to compose my primary one. For example:
```jsonc
{
    // Compose a primary settings configuration from sub-configurations
    "includeConfigs": [
        "/path/to/bobs_configs/editorSettings.json",
        "/path/to/jims_configs/terminalSettings.json"
    ],

    // Keys in this primary file override includes
    "editor.formatOnSave": true
}
```

**Comment by peter2796b** [comment]
This should help
[https://laurens.io/blog/changing-vs-code-settings-locally-when-its-checked-into-git/](https://laurens.io/blog/changing-vs-code-settings-locally-when-its-checked-into-git/)

**Comment by ulzha** [comment]
@peter2796b looks like my issue was that some settings are more special than others. My project was using copilot instructions checked in in .vscode/settings.json. Those are still grayed out and not usable. I tried altering a more mundane setting such as cursor style, and that worked.

![Image](https://github.com/user-attachments/assets/eb1f98a6-7d9e-4bff-a724-103fb2ec49a1)

**Comment by Yzrsah** [comment]
I feel embarrassed to post this but I just wasted 10 hours trying to fix my config files to synchronize the settings across all my profiles. It ruined the whole day.

Please Microsoft, we need this fixed because it's wasting our time doing busy work that adds up to nothing and has to be re-maintained and

**Comment by Yzrsah** [comment]
@sandy081 I feel embarrassed to post this, but I just wasted 10 hours trying to fix all my profiles to manually merge and synchronize my settings across the profiles yet again. It ruined the whole day.

I feel bad about throwing my day in the garbage. I should have just ignored the problem and left it as a mess that I've lost track of.

When is the last time you threw your day in the garbage, Sandy?

Please Microsoft, this is something that is wasting our time.

**Comment by mletterle** [comment]
Please! 🙏🏻

**Comment by roa-nyx** [comment]
Think this would be a neat feature.

I came across it today as I wanted to be able to color a specific workspace so I don't accidentally get it confused with other VS code entries, but I don't want to ignore the workspace settings file nor do I want to commit a change to the theme/colors of the rest of my team.

Hope to see it prioritized soon.

**Comment by Mirsario** [comment]
This feature is absolutely necessary for comfortable VSC usage on NixOS systems, and on Unix systems that make use of the nix-deriving Home Manager program. As those environments generate symlinks to immutable locations to achieve software configuration reproducibility, the usual setup for applications such as this is that one would have two configuration files, the first being an immutable 'declarative' configuration file (symlink pointing to `/nix/store/***`), and a second mutable non-declarative local overrides file meant to be used for experiments and non-permanent changes, with the latter file overriding settings from the former. This is very easy to achieve in Sublime Text for example.

The issue has been open for so many years with no responses from the assignee, is anyone willing to try shooting a pull request for it?

**Comment by mastershadow** [comment]
+1

**Comment by 56789KD** [comment]
#15906

**Comment by recursivezero** [comment]
+1 — I’d really like to see modular settings support in VS Code.


As someone who maintains multiple large projects and uses dozens of extensions, my `.vscode/settings.json` keeps growing endlessly. It becomes very hard to navigate, diff, and review changes — especially when multiple developers update settings related to different tools.

#### Why this feature matters

- **Separation of concerns:** I’d love to group settings logically (e.g., `editor.json`, `formatter.json`, `spellcheck.json`, `python.json`) rather than dumping everything in one big file.  
- **Cleaner version control:** Each developer or team can modify their related configuration file without merge conflicts in the main `settings.json`.  
- **Reusability:** Common settings (e.g., spelling or editor preferences) could be reused across multiple projects by importing them.  
- **Extensibility for extensions:** Many extensions already support external config files (e.g., ESLint, Prettier, cSpell), but VS Code’s own settings system does not. A unified “extends/include” pattern would make extension configuration more consistent.

#### Desired behavior  

Allow something as folllow: 

```json
{
  "$include": [
    "./settings/editor.json",
    "./settings/python.json",
    "./settings/spellcheck.json"
  ],
  "editor.fontSize": 14
}
````

VS Code would recursively merge these files to form the effective settings, similar to `tsconfig.json`’s `"extends"` or ESLint’s `"extends"` property.

#### Benefits

* Better maintainability for complex setups
* Easier onboarding (new developers can understand settings by category)
* Works nicely with Git and shared workspaces
* Consistent configuration patterns across the ecosystem

This would be a huge quality-of-life improvement for anyone who uses VS Code seriously across multiple tech stacks.
Please consider prioritizing it — it would make the settings system much more scalable.

💡

**Comment by georgefst** [comment]
> This feature is absolutely necessary for comfortable VSC usage on NixOS systems

@Mirsario As it happens, I also use NixOS now, but I've wanted this feature even back when I was on a more traditional OS with some simple dotfiles scripts. I just really don't like the idea of mixing together configuration with state. The VSCode settings file contains mostly preferences that I'll only ever set once, but also unfortunately some things which I change several times a day like `diffEditor.renderSideBySide` and `workbench.activityBar.location`.

**Comment by Netail** [comment]
6th highest requested feature btw, but AI seems to have priority 🥀

**Comment by jonmchan** [comment]
> 6th highest requested feature btw, but AI seems to have priority 🥀

Maybe we can get the AI to fix this problem for us. 😆

**Comment by kerrickstaley** [comment]
**Update May 7**
I've been using Layered Settings for the past 2 months and I'm quite happy with it; it solves this issue for me. I'd recommend trying it if you find yourself here!

**Previous comment**
I also recently created an extension to solve this problem 🙂

[Layered Settings](https://marketplace.visualstudio.com/items?itemName=KerrickStaley.layered-settings) addresses all the use-cases people have mentioned:
- Checking your settings into your dotfiles repo, without checking in machine-specific settings or bits of state that shouldn't be tracked.
- Having separate team and user settings in a workspace.
- Having project settings that override team settings in a monorepo with multiple projects (via symlinks).
- Sharding settings into multiple files by theme.

Here's a comparison with Shared Settings Manager based on my understanding:
- Layered Settings supports more use-cases (Shared Settings Manager just supports the use-case of splitting team and user workspace settings).
- Layered Settings is more flexible (Jsonnet is a Turing-complete language, and you can have as many config layers as you want).
- Layered Settings doesn't clobber `settings.json`—manual changes made through the GUI are preserved and take precedence.
- Layered Settings has an external dependency (the jsonnet binary) whereas Shared Settings Manager has no deps.

**Comment by LukasWillin** [comment]
Please Microsoft. Make this happen. I just ran into this creating team-shared launch-configurations. However some settings will be different depending on the users dev environment. And as far as I can see atm, it is simply not possible without sending json files around to other developers and them having to rename things manually.

**Comment by Wiktorro** [comment]
Got here independently when having same issues as above, please Microsoft Do Consider such feature!
Marking @vscode-prebuilds-bot for attention , won't make it more ignored at point of 10 years freezer....

**Comment by orimay** [comment]
@sandy081, I see this issue is assigned to you. Are you the right person to bother?

**Comment by tanishqmanuja** [comment]
People say software development is sorted after AI, yet issues like this remain open xD

**Comment by bytebeast** [comment]
Also, if there's a way to switch settings.json to settings.jsonc, that would be really helpful. Since JSONC supports comments, it makes it much easier to add notes about why certain settings are there, include descriptions, or organize things into sections. It seems like comments are already being ignored, but the editor still complains because it's treating the file as plain JSON instead of JSONC.

**Comment by Wiktorro** [comment]
> It seems like comments are already being ignored,

yep there's not problem in using // comments, at least I don't have any issues , never had since my first edits.
However it might be nice to accept some formats that in their nature handle config addition/overwriting/concat - hope you get the idea - like mentioned above **jsonnet**. maybe there's more options, I didn't know any json-like alternative formats till now. 
Just highlighting topic, feel free to discuss :)

> [Layered Settings](https://marketplace.visualstudio.com/items?itemName=KerrickStaley.layered-settings) addresses all the use-cases people have mentioned:

**Comment by nk-tedo-001** [comment]
My ideal scenario:

1. I have a local `settings.json` that customizes my workspace.
2. We want to have a separate, Git-tracked `settings.json` that everyone on the team can share, while still allowing each developer to reference it from their local `settings.json`.

**Comment by kerrickstaley** [comment]
@nk-tedo-001 @bytebeast @tanishqmanuja @orimay @LukasWillin I would encourage you to try my plugin [Layered Settings](https://marketplace.visualstudio.com/items?itemName=KerrickStaley.layered-settings). This plugin allows you to have multiple settings files that get merged into the final `settings.json`. More details in the [previous comment](https://github.com/microsoft/vscode/issues/15909#issuecomment-3996196461) (re-posting because it's now too far up that it doesn't automatically load). Hope this is helpful!
